src/transformers/models/llama/modeling_llama.py

```diff
@@ -604,6 +604,39 @@ class LlamaSdpaAttention(LlamaAttention):
     SDPA API.
     """
 
+    def _scaled_dot_product_attention(
+        self,
+        query_states: torch.Tensor,
+        key_states: torch.Tensor,
+        value_states: torch.Tensor,
+        causal_mask: torch.Tensor,
+        q_len: int,
+    ):
+        # This method is abstracted in order to enable Copied from statements for it.
+
+        if torch._dynamo.is_compiling():
+            # Dynamo is unable to capture `self.is_causal and attention_mask is None and q_len > 1` with `fullgraph=True`, hence the need to always
+            # pass `attn_mask=causal_mask` in this case. Ideally, if available, should use `torch._dynamo.is_fullgraph_compiling()`.
+            attn_output = torch.nn.functional.scaled_dot_product_attention(
+                query_states,
+                key_states,
+                value_states,
+                attn_mask=causal_mask,
+                dropout_p=self.attention_dropout if self.training else 0.0,
+            )
+        else:
+            # In case we are not compiling, we may set `causal_mask` to None, which is required to dispatch to SDPA's Flash Attention 2 backend, rather
+            # relying on the `is_causal` argument.
+            attn_output = torch.nn.functional.scaled_dot_product_attention(
+                query_states,
+                key_states,
+                value_states,
+                attn_mask=causal_mask,
+                dropout_p=self.attention_dropout if self.training else 0.0,
+                is_causal=self.is_causal and causal_mask is None and q_len > 1,
+            )
+        return attn_output
```

Comment:
Some quick testing seems to confirm this:

> When I train a compiled model without passing an attention_mask to LlamaForCausalLM, the if torch._dynamo.is_compiling() statement appears to apply Bidirectional Attention instead of Causal Attention (since is_causal defaults to False in scaled_dot_product_attention and causal_mask is None).

Here's the allclose comparison between `LlamaFlashAttention2`, eager `LlamaSdpaAttention` and compiled `LlamaSdpaAttention` in the script below.

```text
Flash vs Eager SDPA allclose: True
Flash vs Compiled SDAP allclose: False
Eager SDPA vs Compiled SDAP allclose: False
```

<details>

```python
import torch
from transformers.models.llama.modeling_llama import LlamaConfig, LlamaFlashAttention2, LlamaSdpaAttention
torch.set_float32_matmul_precision('high')

config = LlamaConfig(hidden_size=2048, num_attention_heads=16)
sdpa_attention = LlamaSdpaAttention(config, layer_idx=1).to(dtype=torch.bfloat16, device='cuda')
flash_attention = LlamaFlashAttention2(config, layer_idx=1)

flash_attention.load_state_dict(sdpa_attention.state_dict())
flash_attention.to(dtype=torch.bfloat16, device='cuda')

input_tensor = torch.randn(1, 16, 2048, dtype=torch.bfloat16, device='cuda')
position_ids = torch.arange(0, 16, dtype=torch.int64, device='cuda').unsqueeze(0)

compiled_sdpa_attention = torch.compile(sdpa_attention)

with torch.no_grad():
    out_sdpa = sdpa_attention(input_tensor, position_ids=position_ids)
    out_compiled_sdpa = compiled_sdpa_attention(input_tensor, position_ids=position_ids)
    out_flash = flash_attention(input_tensor, position_ids=position_ids)

print(f"Flash vs Eager SDPA allclose: {torch.allclose(out_flash[0], out_sdpa[0], atol=1e-3, rtol=1e-2)}")
print(f"Flash vs Compiled SDAP allclose: {torch.allclose(out_flash[0], out_compiled_sdpa[0], atol=1e-3, rtol=1e-2)}")
print(f"Eager SDPA vs Compiled SDAP allclose: {torch.allclose(out_sdpa[0], out_compiled_sdpa[0], atol=1e-3, rtol=1e-2)}")
```

</details>

Reply:
@warner-benjamin Thank you! Regarding your script, I believe LlamaAttention is not meant to be used in isolation, for example currently it depends a lot on this other method: https://github.com/huggingface/transformers/blob/1fc34aa6663744d1ef53674b560b0f2ca061e1db/src/transformers/models/llama/modeling_llama.py#L1060 & on passing the `attention_mask` input.